src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ActiveDirectoryAuthenticationProvider.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Security;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+
+namespace Microsoft.Data.SqlClient
+{
+
+    /// <summary>
+    /// Default auth provider for AD Integrated.
+    /// </summary>
+    internal class ActiveDirectoryAuthenticationProvider : SqlAuthenticationProvider
+    {
+        private static readonly string s_defaultScopeSuffix = "/.default";
+        private readonly string _type = typeof(ActiveDirectoryAuthenticationProvider).Name;
+        private readonly SqlClientLogger _logger = new SqlClientLogger();
+
+        /// <summary>
+        /// Get token.
+        /// </summary>
+        public override Task<SqlAuthenticationToken> AcquireTokenAsync(SqlAuthenticationParameters parameters) => Task.Run(async () =>
+        {
+            AuthenticationResult result;
+            string scope = parameters.Resource.EndsWith(s_defaultScopeSuffix) ? parameters.Resource : parameters.Resource + s_defaultScopeSuffix;
+            string[] scopes = new string[] { scope };
+
+            if (parameters.AuthenticationMethod == SqlAuthenticationMethod.ActiveDirectoryServicePrincipal)
+            {
+                IConfidentialClientApplication ccApp = ConfidentialClientApplicationBuilder.Create(parameters.UserId)
+                    .WithAuthority(parameters.Authority)
+                    .WithClientSecret(parameters.Password)
+                    .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                    .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                    .Build();
+
+                result = ccApp.AcquireTokenForClient(scopes).ExecuteAsync().Result;
+                return new SqlAuthenticationToken(result.AccessToken, result.ExpiresOn);
+            }
+
+            IPublicClientApplication app = PublicClientApplicationBuilder.Create(ActiveDirectoryAuthentication.AdoClientId)
+                .WithAuthority(parameters.Authority)
+                .WithClientName(Common.DbConnectionStringDefaults.ApplicationName)
+                .WithClientVersion(Common.ADP.GetAssemblyVersion().ToString())
+                .WithRedirectUri("http://localhost")
```

Comment:
The localhost redirect URI is only applicable to .NET Core but does not hurt to have it for all platforms. We will need to add it to our registered first party application for it to start working. We may want to consider getting our own app ID for Microsoft.Data.SqlClient instead of continuing to use the one for System.Data.SqlClient.